In [19]:
import numpy as np
import pandas as pd
pd.__version__

'2.1.4'

Creating a DataFrame

In [12]:
classes = pd.Series(['Maths', 'Chemistry', 'Physics', 'History', 'Geography'])
grades = pd.Series([90, 54, 77, 22, 25])
dataframe = pd.DataFrame({'Subjects':classes, 'Grades': grades})
dataframe.head()

Subjects  Grades
0      Maths      90
1  Chemistry      54
2    Physics      77
3    History      22
4  Geography      25

Load the dataframe

In [14]:
dataframe = pd.read_csv('data/black_friday_sale.csv', 
                        index_col=0 # indicates first columns as the index
                        )

In [15]:
# Just show us the first rows in the dataframe
dataframe.head(n=10) # n: represent the number of columns 

Product_ID Gender    Age  ...  Product_Category_1 Product_Category_2 Product_Category_3
User_ID                           ...                                                          
1000004  P00128942      M  46-50  ...                   1               11.0                NaN
1000009  P00113442      M  26-35  ...                   3                5.0                NaN
1000010  P00288442      F  36-45  ...                   5               14.0                NaN
1000010  P00145342      F  36-45  ...                   4                9.0                NaN
1000011  P00053842      F  26-35  ...                   4                5.0               12.0
1000013  P00350442      M  46-50  ...                   2                3.0               15.0
1000013  P00155442      M  46-50  ...                   1               11.0               15.0
1000013   P0094542      M  46-50  ...                   2                4.0                9.0
1000015  P00161842      M  26-35  ...                  10               13.0               16.0
1000022  P00067942      M  18-25  ...                   5               14.0                NaN

[10 rows x 10 columns]

The method `describe()` just works for numerical values, the rest are ignored.

In [16]:
dataframe.describe()

Occupation  Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3
count  233599.000000   233599.000000       233599.000000       161255.000000        71037.000000
mean        8.085407        0.410070            5.276542            9.849586           12.669454
std         6.521146        0.491847            3.736380            5.094943            4.125944
min         0.000000        0.000000            1.000000            2.000000            3.000000
25%         2.000000        0.000000            1.000000            5.000000            9.000000
50%         7.000000        0.000000            5.000000            9.000000           14.000000
75%        14.000000        1.000000            8.000000           15.000000           16.000000
max        20.000000        1.000000           18.000000           18.000000           18.000000

Now we will clean the data. We could have:

- Empty cells
- Data in wrong format
- Wrong data
- Duplicates

In [17]:
# Remove empty cells
new_dataframe = dataframe.dropna()
new_dataframe.describe()

Occupation  Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3
count  71037.000000    71037.000000        71037.000000        71037.000000        71037.000000
mean       8.219801        0.402002            2.725185            6.882287           12.669454
std        6.482756        0.490306            2.564070            4.510544            4.125944
min        0.000000        0.000000            1.000000            2.000000            3.000000
25%        3.000000        0.000000            1.000000            2.000000            9.000000
50%        7.000000        0.000000            1.000000            6.000000           14.000000
75%       14.000000        1.000000            4.000000           10.000000           16.000000
max       20.000000        1.000000           15.000000           16.000000           18.000000

Values from `Occupation`, `Marital_Status`, `Product_Category_1` and `Product_Category_2` were removed, leaving all numerical columns with 71037 values.

Nevertheless, we could fill the `nan_values` with other statistics.

In [18]:
# Fill nan_values with zero
new_dataframe = dataframe.fillna(value=0)
new_dataframe.describe()

Occupation  Marital_Status  Product_Category_1  Product_Category_2  Product_Category_3
count  233599.000000   233599.000000       233599.000000       233599.000000       233599.000000
mean        8.085407        0.410070            5.276542            6.799237            3.852756
std         6.521146        0.491847            3.736380            6.217668            6.256634
min         0.000000        0.000000            1.000000            0.000000            0.000000
25%         2.000000        0.000000            1.000000            0.000000            0.000000
50%         7.000000        0.000000            5.000000            5.000000            0.000000
75%        14.000000        1.000000            8.000000           14.000000            8.000000
max        20.000000        1.000000           18.000000           18.000000           18.000000

In [28]:
dataframe['Product_Category_1'].replace(to_replace=np.nan, value=0, inplace=True)
dataframe['Product_Category_2'].replace(to_replace=np.nan, value=0, inplace=True)
dataframe['Product_Category_3'].replace(to_replace=np.nan, value=0, inplace=True)

dataframe['Stay_In_Current_City_Years'].replace(to_replace='4+', value='4', inplace=True)

In [29]:
dataframe.head()

Product_ID Gender Age  ...  Product_Category_1 Product_Category_2 Product_Category_3
User_ID                        ...                                                          
1000004  P00128942      M  50  ...                   1               11.0                0.0
1000009  P00113442      M  35  ...                   3                5.0                0.0
1000010  P00288442      F  45  ...                   5               14.0                0.0
1000010  P00145342      F  45  ...                   4                9.0                0.0
1000011  P00053842      F  35  ...                   4                5.0               12.0

[5 rows x 10 columns]

In [30]:
# Obtain the unique values
dataframe.Age.unique()

array(['50', '35', '45', '25', '55', '60', '17'], dtype=object)

In [31]:
dataframe.Age.replace(to_replace=['46-50', '26-35', '36-45', '18-25', '51-55', '55+', '0-17'],
                      value=['50', '35', '45', '25', '55', '60', '17'],
                      inplace=True)
dataframe.head()

Product_ID Gender Age  ...  Product_Category_1 Product_Category_2 Product_Category_3
User_ID                        ...                                                          
1000004  P00128942      M  50  ...                   1               11.0                0.0
1000009  P00113442      M  35  ...                   3                5.0                0.0
1000010  P00288442      F  45  ...                   5               14.0                0.0
1000010  P00145342      F  45  ...                   4                9.0                0.0
1000011  P00053842      F  35  ...                   4                5.0               12.0

[5 rows x 10 columns]

The method `.dtypes` shows us the types of each column in the dataset.

In [33]:
dataframe.dtypes

Product_ID                     object
Gender                         object
Age                            object
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years     object
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
dtype: object

Now we want convert some columns in the correct `dtype`

In [35]:
dataframe.Age = dataframe.Age.astype('int')
dataframe.Stay_In_Current_City_Years  = dataframe.Stay_In_Current_City_Years .astype('int')
dataframe.dtypes

Product_ID                     object
Gender                         object
Age                             int64
Occupation                      int64
City_Category                  object
Stay_In_Current_City_Years      int64
Marital_Status                  int64
Product_Category_1              int64
Product_Category_2            float64
Product_Category_3            float64
dtype: object

Now we could use `drop_duplicate()` to eliminate duplicates values in a row or columns

In [39]:
dataframe.drop_duplicates(subset='Product_ID', keep='first', inplace=True, ignore_index=True)
dataframe.head()

Product_ID Gender  Age  ...  Product_Category_1 Product_Category_2  Product_Category_3
0  P00128942      M   50  ...                   1               11.0                 0.0
1  P00113442      M   35  ...                   3                5.0                 0.0
2  P00288442      F   45  ...                   5               14.0                 0.0
3  P00145342      F   45  ...                   4                9.0                 0.0
4  P00053842      F   35  ...                   4                5.0                12.0

[5 rows x 10 columns]

We even use pandas to obtain the correlation between variables using `.corr()`

In [38]:
dataframe.corr()

ValueError: could not convert string to float: 'P00128942'